### PR TITLE
ChatX module

### PR DESCRIPTION
**Pull request by WilliamChelman**
A new route has been added for a small chat module. Here are the steps to enjoy this wonderful module:

1. Clone this repo
2. Do a `npm install` to install dependencies
3. Run `npm start -s` to start a local server, a page should automatically open in your favorite browser at localhost:3000
4. First you have to login and/or create an account, because the new route is only accessible to logged users
5. When logged, click on the "Chat Rooms" link on top of the page
6. On the left side you will see a list of available chats in "Other chats" that you can join. You can also create your own chat using the "Create new chat" input
7. When you join or create a chat, it is added in the "My chats" list. Here you can either "see" a chat or "leave" it. Also a list of the users in the active chat is displayed on the bottom left of the screen.
8. When you leave a chat it will come back to the "Other chats' list
9. When you "see" a chat, you will see the 10 latest messages and you can add new messages by filling the input at the bottom of the screen

### CODE REVIEW

**Comment by alexstrat** [comment]
Can't do `npm start -s`:
```txt
ERROR in ./src/components/chatRooms/ChatRoomForm.js
Module not found: Error: Cannot resolve module 'react/lib/ReactBaseClasses' in /Users/alexandrelacheze/Dev/react-hot-redux-firebase-starter/src/components/chatRooms
 @ ./src/components/chatRooms/ChatRoomForm.js 17:24-61

ERROR in ./src/components/chatRooms/MessageForm.js
Module not found: Error: Cannot resolve module 'react/lib/ReactBaseClasses' in /Users/alexandrelacheze/Dev/react-hot-redux-firebase-starter/src/components/chatRooms
 @ ./src/components/chatRooms/MessageForm.js 17:24-61
```

Any idea?

File: src/components/chatRooms/ChatRoomForm.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      }
+    };
+
+    this.save = this.save.bind(this);
+    this.change = this.change.bind(this);
+  }
+
+  save(event) {
+    event.preventDefault();
+    ChatRoomApi.create(this.state.chatRoom);
```
Comment (alexstrat) [inline src/components/chatRooms/ChatRoomForm.js:21]: This is not very redux-friendly: in redux, everything should happen thru [actions](https://redux.js.org/docs/basics/Actions.html) and some components (the [containers](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0)) are [connected](https://redux.js.org/docs/basics/UsageWithReact.html) to actions.

**Comment by alexstrat** [comment]
```sh
$ rm -rf node_modules
$ npm i
```
did the job for me.

File: src/components/chatRooms/ChatRoomForm.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    const chatRoom = this.state.chatRoom;
+    chatRoom.name = '';
```
Comment (alexstrat) [inline src/components/chatRooms/ChatRoomForm.js]: ⚠️ mutate data in react is very [not-recommanded](https://medium.com/pro-react/a-brief-talk-about-immutability-and-react-s-helpers-70919ab8ae7c).

File: src/components/chatRooms/ChatRoomForm.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  isDisabled() {
+    return !this.state.chatRoom.name.trim();
+  }
+
+  change(event) {
+    const field = event.target.name;
+    let chatRoom = this.state.chatRoom;
+    chatRoom[field] = event.target.value;
```
Comment (alexstrat) [inline src/components/chatRooms/ChatRoomForm.js]: ⚠️ state mutation

File: src/components/chatRooms/ChatRoomForm.js (hunk 2 above)
Comment (WilliamChelman) [reply src/components/chatRooms/ChatRoomForm.js]: So instead, I should do something like this right ?
```javascript
const chatRoom = {name: ''};
this.setState({chatRoom});
```
And so never change directly something referenced in the `this.state` object ?

File: src/components/chatRooms/ChatRoom.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  onMessageAddition(data) {
+    const messages = this.state.messages;
+    const newMessage = data.val();
+    newMessage.key = data.key;
+    messages.push(newMessage);
+    this.setState({messages});
+  }
+
+  getClassName(message) {
+    let className = 'col-md-8';
```
Comment (alexstrat) [inline src/components/chatRooms/ChatRoom.js]: https://github.com/JedWatson/classnames

File: src/components/chatRooms/ChatRoom.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      this.registerToMessages(nextProps.chat.key);
+    }
+  }
+
+  componentWillUnmount() {
+    this.unregisterToMessages();
+  }
+
+  unregisterToMessages() {
+    MessageApi.getRef().off('child_added', this.onMessageAddition);
```
Comment (alexstrat) [inline src/components/chatRooms/ChatRoom.js:30]: Not using redux?

File: src/components/chatRooms/ChatRoomForm.js (hunk 2 above)
Comment (alexstrat) [reply src/components/chatRooms/ChatRoomForm.js]: Yes!

File: src/components/chatRooms/ChatRoomForm.js (hunk 1 above)
Comment (alexstrat) [reply src/components/chatRooms/ChatRoomForm.js:21]: With redux, you should not use API directly but use actions instead

File: src/components/chatRooms/MessageForm.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+function mapStateToProps(state, ownProps) {
+  return {
+    activeChat: state.chat.activeChat
+  };
+}
+
+function mapDispatchToProps(dispatch) {
+  return {
+    actions: bindActionCreators({createMessage}, dispatch)
```
Comment (alexstrat) [inline src/components/chatRooms/MessageForm.js:74]: There is lot a components that are connected, does not follow the container // component recommendation: https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0

**Pull request closed without merging**